## A "Hello World" Example of Machine Learning - Revisit

Loading the Iris dataset from scikit-learn. 

The first column represents Sepal length, the second column represents Sepal width,  the third column represents the petal length, and the fourth column the petal width of the flower samples. The classes (type of species) are already converted to integer labels where 0=Iris-Setosa, 1=Iris-Versicolor, 2=Iris-Virginica.

Here, we are using only two features: the third and fourth columns. 

In [1]:
from sklearn import datasets
import numpy as np

iris = datasets.load_iris()
#iris.data

In [2]:
X = iris.data[:, [0, 3]]


In [3]:
y = iris.target

In [4]:
print('Class labels:', np.unique(y))

Class labels: [0 1 2]


Scikit-learn algorithms support multi-class classification via the One-Versus-Rest(OvR) method. 

Splitting data into 70% training and 30% test data:

In [5]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.3, random_state=1, stratify=y)

In [6]:
print('Labels counts in y:', np.bincount(y))
print('Labels counts in y_train:', np.bincount(y_train))
print('Labels counts in y_test:', np.bincount(y_test))

Labels counts in y: [50 50 50]
Labels counts in y_train: [35 35 35]
Labels counts in y_test: [15 15 15]


### Standardizing the features:

In [7]:
from sklearn.preprocessing import StandardScaler

sc = StandardScaler() #center the distribution around zero (mean), with a standard deviation of 1.
sc.fit(X_train)
X_train_std = sc.transform(X_train)
X_test_std = sc.transform(X_test)

In [8]:
from sklearn.linear_model import Perceptron

ppn = Perceptron(max_iter=100, eta0=0.1, random_state=42)
ppn.fit(X_train_std, y_train)

Perceptron(eta0=0.1, max_iter=100, random_state=42)

### Test the model with the hold-out test set

In [9]:
y_pred = ppn.predict(X_test_std)
print('Misclassified samples: ' + str((y_test != y_pred).sum()))

Misclassified samples: 10


In [10]:
from sklearn.metrics import accuracy_score

print('Accuracy: ' + str(accuracy_score(y_test, y_pred)))

Accuracy: 0.7777777777777778


In [11]:
X_new = [[1.1, 0.2],[0.4, 1.9], [1.4, 0.2]]
y_new = ppn.predict(X_new)
y_new

array([2, 2, 2])

### Evaluate the model using cross validation

In [12]:
from sklearn.model_selection import cross_val_score
cross_val_score(ppn, X_train_std, y_train, cv=4, scoring="accuracy")

array([0.88888889, 0.76923077, 0.53846154, 0.80769231])

2-features: sccuracy score: array([0.88888889, 0.48148148, 0.7037037 , 0.95833333])

### Exercise 1: Use all four features to train the model and use cross validaton to check if the results better? Briefly explain why. 

In [18]:
X1 = iris.data[:]
X1_train, X1_test, y1_train, y1_test = train_test_split(
    X1, y, test_size=0.3, random_state=1, stratify=y)

sc.fit(X1_train)
X1_train_std = sc.transform(X1_train)
X1_test_std = sc.transform(X1_test)

ppn.fit(X1_train_std, y1_train)
y1_pred = ppn.predict(X1_test_std)
print('Misclassified samples: ' + str((y1_test != y1_pred).sum()))
print('Accuracy: ' + str(accuracy_score(y1_test, y1_pred)))

cross_val_score(ppn, X1_train_std, y1_train, cv=4, scoring="accuracy")

Misclassified samples: 3
Accuracy: 0.9333333333333333


array([0.92592593, 0.88461538, 0.84615385, 0.88461538])

### Exercise 2: Try with the scikit-learn stochastic gradient descent model instead of perceptron. Use all four features. Evaluate with cross-validation how does the model perform in terms of accuracy using both two features and four features. 

In [20]:
from sklearn.linear_model import SGDClassifier
sgd = SGDClassifier(random_state=42)

sgd.fit(X_train_std, y_train)
y_pred = sgd.predict(X_test_std)
print('Misclassified samples: ' + str((y_test != y_pred).sum()))
print('Accuracy: ' + str(accuracy_score(y_test, y_pred)))

sgd.fit(X1_train_std, y1_train)
y_pred = sgd.predict(X1_test_std)
print('Misclassified samples: ' + str((y1_test != y_pred).sum()))
print('Accuracy: ' + str(accuracy_score(y1_test, y_pred)))

Misclassified samples: 1
Accuracy: 0.9777777777777777
Misclassified samples: 5
Accuracy: 0.8888888888888888


In [ ]:
#With the perceptron model, switching from 2 to 4 features has improved the model accuracy, but with the stochastic gradient descent model the accuracy was better using 2 features instead of 4 features